ADR/0043-artifact-constructors.md

```diff
@@ -0,0 +1,104 @@
+These pipelines would be executed after a Snapshot is generated and before the integration test pipelines start.
+
+When each artifact generation pipelineRun is completed, the reference to the artifact is extracted and added to the original Snapshot.
+After that, the integration testing pipelines castart, and since the artifacts are referenced by the Snapshot,
+it is possible to fetch them and run tests on them.
+
+```yaml
+apiVersion: appstudio.redhat.com/v2alpha1
+kind: ArtifactConstructor
+metadata:
+    name: example--artifact-constructor
+spec:
+    selectors:
+        - expression: ‘metadata.labels["konflux-ci.dev/application"] == "integration-service" && metadata.annotations["test.pac.appstudio.openshift.io/event-type"] == "pull_request"’
+          description: “Selects only PR Snapshots associated with integration-service”
+    params:
+      - name: source
+        value: https://github.com/konflux-ci/integration-service
+    resolverRef:
+      resolver: git
+      params:
+        - name: url
+          value: https://github.com/konflux-ci/integration-examples
+        - name: revision
+          value: main
+        - name: pathInRepo
+          value: pipelines/bundle_generator_pipeline.yaml
+    extractors:
+      - ref: 'status.results.filter(result,result.name == "IMAGE_URL")[0].value'
+        name: ‘metadata.labels["appstudio.openshift.io/component"]’
+
+```
+
+Don't take the fields literally. The details will likely change after the ADR has settled. It is
+here for illustration purposes.
+
+The `selector` field will be used to identify resources that should trigger the construction of new Artifacts. This will be done using CEL expressions.
+
+The `resolverRef` resource will be used to define the Tekton pipeline definition which can be used to run the artifact pipelineRun.
+
+The `extractor` resource will use CEL expressions to extract the individual field values from the artifact pipelineRun.
+The individual artifacts are expected to use a reference that can be used to fetch it, like an image pullSpec or similar URL.
+Since the `extractor` resource is a list, it can take multiple extractor entries if the artifact pipeline is expected to provide multiple artifacts.
+Each extracted resource will be added to the Snapshot's metadata.
+
+## Consequences
+
+- [integration-service] now executes additional Tekton pipelineRuns which are expected to generate artifacts which
+  can then be associated with the Snapshot.
+- [integration-service] waits for the artifact pipelineRuns to finish before starting the integration testing pipelines
+  for a given Snapshot.
+- this will have implications for the [release service] as it will further remove coupling around an Application and an Application Snapshot
+- Potential decrease of a reliance on an Application, which will have implications on [build-service] and the Konflux UI
```

Comment:
Build Service does not rely on Application, main resource is a Component. From Build Service point of view, we can delete Application CR even today.

Reply:
Good point, I think that a better term would be the `Application/Component model`.